File: 20. Valid Parentheses/py/ValidParentheses.py
# LeetCode problem 20. Valid Parentheses
# https://leetcode.com/problems/valid-parentheses/description/

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        dirty = False
        for i in s:
            if i in ['(', '[', '{'] or len(stack) == 0:
                stack.append(i)
                dirty = True
            elif len(stack) > 0:
                sub = stack[-1] + i
                if sub in ['()', '[]', '{}']:
                    stack.pop()
                else:
                    return False
        if len(stack) == 0 and dirty:
            return True
        else:
            return False

File: 20. Valid Parentheses/cs/ValidParentheses/ValidParentheses.cs
// LeetCode problem: 20. Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/description/

namespace LeetCode.ValidParentheses;

public class Solution
{
    public bool IsValid(string s)
    {
        var stack = new Stack<char>();
        var pairs = new Dictionary<char, char>
        {
            [')'] = '(',
            [']'] = '[',
            ['}'] = '{'
        };

        foreach (char c in s)
        {
            if (c is '(' or '[' or '{')
                stack.Push(c);
            else if (stack.Count == 0 || stack.Peek() != pairs[c])
                return false;
            else
                stack.Pop();
        }

        return s.Length > 0 && stack.Count == 0;
    }
}
